util/election/runner.go

```diff
@@ -0,0 +1,187 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package election
+
+import (
+	"context"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian/util"
+)
+
+const (
+	// Minimum values for configuration intervals.
+	MinPreElectionPause    = 10 * time.Millisecond
+	MinMasterCheckInterval = 50 * time.Millisecond
+	MinMasterHoldInterval  = 10 * time.Second
+)
+
+// RunnerConfig describes the parameters for an election Runner.
+type RunnerConfig struct {
+	// PreElectionPause is the maximum interval to wait before starting a
+	// mastership election for a particular log.
+	PreElectionPause time.Duration
+	// MasterCheckInterval is the interval between checks that we still
+	// hold mastership for a log.
+	MasterCheckInterval time.Duration
+	// MasterHoldInterval is the minimum interval to hold mastership for.
+	MasterHoldInterval time.Duration
+	// ResignOdds gives the chance of resigning mastership after each
+	// check interval, as the N for 1-in-N.
+	ResignOdds int
+
+	TimeSource util.TimeSource
+}
+
+// fixupRunnerConfig ensures operation parameters have required minimum values.
+func fixupRunnerConfig(cfg *RunnerConfig) {
+	if cfg.PreElectionPause < MinPreElectionPause {
+		cfg.PreElectionPause = MinPreElectionPause
+	}
+	if cfg.MasterCheckInterval < MinMasterCheckInterval {
+		cfg.MasterCheckInterval = MinMasterCheckInterval
+	}
+	if cfg.MasterHoldInterval < MinMasterHoldInterval {
+		cfg.MasterHoldInterval = MinMasterHoldInterval
+	}
+	if cfg.ResignOdds < 1 {
+		cfg.ResignOdds = 1
+	}
+	if cfg.TimeSource == nil {
+		cfg.TimeSource = util.SystemTimeSource{}
+	}
+}
+
+// Runner controls a continuous election process.
+type Runner struct {
+	// Allow the user to store a Cancel function with the runner for convenience.
+	Cancel   context.CancelFunc
+	id       int64
+	cfg      *RunnerConfig
+	tracker  *MasterTracker
+	election MasterElection
+}
+
+// NewRunner builds a new election Runner instance with the given configuration.  On calling
+// Run(), the provided election will be continuously monitored and mastership changes will
+// be notified to the provided MasterTracker instance.
+func NewRunner(id int64, cfg *RunnerConfig, tracker *MasterTracker, cancel context.CancelFunc, el MasterElection) *Runner {
+	fixupRunnerConfig(cfg)
+	return &Runner{
+		Cancel:   cancel,
+		id:       id,
+		cfg:      cfg,
+		tracker:  tracker,
+		election: el,
+	}
+}
+
+// Run performs the process of performing a continuous election. It runs continuously until
```

Comment:
Could you rephrase it to avoid "performs performing"? E.g.:
```
// Run continuously takes part in master election on behalf of this instance.
// Returns when the context is canceled or an internal error is encountered.
```

Reply:
Done.